### PR TITLE
Support new build methods

### PR DESCRIPTION
**Pull request by gpotter2**
The goals of this PR is to support `python -m build` to generate wheels along the sdist.

This PR:
- makes sure that `scapy/VERSION` is also exported in wheels
- update the documentation to reflect the new build methods: `python -m build`, and the new 3.7+ dependencies. Also updated some Windows parts and dependency to `tcpdump` (which is now to `libpcap`).
- includes https://github.com/secdev/scapy/pull/3945

closes https://github.com/secdev/scapy/pull/3945
closes https://github.com/secdev/scapy/issues/3955

### CODE REVIEW

**Comment by evverx** [comment]
Thanks!

I can't say I fully understand how it all works (though I hope I'm slowly getting there :-)) but it seems it isn't possible to enforce versions downstream using something like
```
SCAPY_VERSION=2.10.0 python3 -m build
...
Successfully built scapy-2.5.1.dev68.tar.gz and scapy-2.5.1.dev68-py3-none-any.whl
...
(cd ..; python3 -c 'import scapy; print(scapy.VERSION)')
2.5.1.dev68
```
My understanding is that `SCAPY_VERSION` should take precedence over `git describe`.

**Comment by evverx** [comment]
> removes the VERSION file

With `scapy/VERSION` gone at least Debian and Gentoo would have to either switch to the `SCAPY_VERSION` environment variable or just rely on the archive tag. It isn't the end of world probably but I think it would be great to mention it in the release notes when the next tarball is released.

FWIW I generated the git archive and when I tried to unpacked it I got a lot of warnings like
```
...
tox.ini
tar: test: time stamp 2023-03-26 23:56:41 is 34702.105210495 s in the future
tar: tox.ini: time stamp 2023-03-26 23:56:41 is 34702.105116622 s in the future
```
and when I tried to build scapy from inside the unpacked tarball it failed with:
<details>
  <summary>`ERROR Backend subprocess exited when trying to invoke get_requires_for_build_sdist`</summary>

```
* Creating virtualenv isolated environment...
* Installing packages in isolated environment... (setuptools>=62.0.0, setuptools_scm[toml]>=6.2)
* Getting build dependencies for sdist...
Traceback (most recent call last):
  File "/VENV/lib/python3.11/site-packages/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
    main()
  File "/VENV/lib/python3.11/site-packages/pyproject_hooks/_in_process/_in_process.py", line 335, in main
    json_out['return_val'] = hook(**hook_input['kwargs'])
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/VENV/lib/python3.11/site-packages/pyproject_hooks/_in_process/_in_process.py", line 287, in get_requires_for_build_sdist
    return hook(config_settings)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/build-env-54sucmvl/lib/python3.11/site-packages/setuptools/build_meta.py", line 341, in get_requires_for_build_sdist
    return self._get_build_requires(config_settings, requirements=[])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/build-env-54sucmvl/lib/python3.11/site-packages/setuptools/build_meta.py", line 320, in _get_build_requires
    self.run_setup()
  File "/tmp/build-env-54sucmvl/lib/python3.11/site-packages/setuptools/build_meta.py", line 335, in run_setup
    exec(code, locals())
  File "<string>", line 38, in <module>
  File "/tmp/build-env-54sucmvl/lib/python3.11/site-packages/setuptools/__init__.py", line 108, in setup
    return distutils.core.setup(**attrs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/build-env-54sucmvl/lib/python3.11/site-packages/setuptools/_distutils/core.py", line 147, in setup
    _setup_distribution = dist = klass(attrs)
                                 ^^^^^^^^^^^^
  File "/tmp/build-env-54sucmvl/lib/python3.11/site-packages/setuptools/dist.py", line 488, in __init__
    _Distribution.__init__(
  File "/tmp/build-env-54sucmvl/lib/python3.11/site-packages/setuptools/_distutils/dist.py", line 283, in __init__
    self.finalize_options()
  File "/tmp/build-env-54sucmvl/lib/python3.11/site-packages/setuptools/dist.py", line 912, in finalize_options
    ep(self)
  File "/tmp/build-env-54sucmvl/lib/python3.11/site-packages/setuptools_scm/integration.py", line 127, in infer_version
    _assign_version(dist, config)
  File "/tmp/build-env-54sucmvl/lib/python3.11/site-packages/setuptools_scm/integration.py", line 63, in _assign_version
    _version_missing(config)
  File "/tmp/build-env-54sucmvl/lib/python3.11/site-packages/setuptools_scm/__init__.py", line 108, in _version_missing
    raise LookupError(
LookupError: setuptools-scm was unable to detect version for /TARBALL.

Make sure you're either building from a fully intact git repository or PyPI tarballs. Most other sources (such as GitHub's tarballs, a git checkout without the .git folder) don't contain the necessary metadata and will not work.

For example, if you're using pip, instead of https://github.com/user/proj/archive/master.zip use git+https://github.com/user/proj.git#egg=proj

ERROR Backend subprocess exited when trying to invoke get_requires_for_build_sdist
```
</details>

**Comment by gpotter2** [comment]
> When I tried to build scapy from inside the unpacked tarball it failed with

Hmm, that sucks. I'm suddenly really unsure about this PR, we don't want to break tarballs. I'll check again

File: setup.py [hunk 1]
```diff
@@ -37,19 +38,51 @@ def process_ignore_tags(buffer):
...
+
+
+class BuildPy(build_py):
+    """
+    Modified build_py to create scapy/VERSION file
+    """
+    def build_package_data(self):
+        super(BuildPy, self).build_package_data()
+        # ensure there's a scapy/VERSION file
+        _build_version(self.build_lib)
```
Comment (evverx) [inline setup.py:82]: I think now that `scapy/VERSION` is added to the sdist and wheel automatically it can be removed from `MANIFEST.in` to prevent `setuptools` from complaining about it:
```sh
warning: no files found matching 'scapy/VERSION'
```


FWIW there seems to be another warning when the wheel is generated from the sdist:
```sh
warning: no previously-included files matching '*' found under directory 'test'
```
but it can be reproduced with the master branch as well so it's probably a separate issue.

File: tox.ini [hunk 2]
```diff
@@ -139,10 +139,11 @@ commands = codespell --ignore-words=.config/codespell_ignore.txt --skip="*.pyc,*
...
 skip_install = true
 deps = twine
-       setuptools>=38.6.0
        cmarkgfm
-commands = python setup.py --quiet sdist
-           twine check dist/*
+       build
+setenv = SCAPY_VERSION=3.0.0
+commands = python -m build
+           twine check --strict dist/*
```
Comment (evverx) [inline tox.ini:146]: I think it would be great to eventually add something like
```sh
git archive HEAD >scapy.tar
python3 -m pip install scapy.tar # or just unpack it and run `python3 -m build` inside the directory instead
```
just to make sure that tarballs are buildable and installable. It can be done later though.

File: setup.py (hunk 1 above)
Comment (evverx) [reply setup.py:82]: > FWIW there seems to be another warning when the wheel is generated from the sdist:

It was introduced in 6ae0e87fba24e9e61eaf687d5374a5dbd330088d as far as I can see.

Comment (gpotter2) [reply setup.py:82]: The two comments appear to be correct !

File: tox.ini (hunk 2 above)
Comment (gpotter2) [reply tox.ini:146]: No you're right, let's do this now to prevent regressions.

File: MANIFEST.in [hunk 3]
```diff
@@ -1,5 +1,3 @@
 include MANIFEST.in
 include LICENSE
 include run_scapy
-include scapy/VERSION
-recursive-exclude test *
```
Comment (evverx) [inline MANIFEST.in:5]: Looks like the warnings are gone but the test directory seems to end up in the sdist somehow even though `pyproject.toml` explicitly excludes it:
```sh
$ tar tzvf dist/scapy-2.5.0.dev72.tar.gz '*/test'
drwxr-xr-x root/root         0 2023-04-09 13:07 scapy-2.5.0.dev72/test/
-rw-rw-r-- root/root      5126 2023-04-08 20:52 scapy-2.5.0.dev72/test/testsocket.py
```
I would have expected it to somehow end up in the wheel too but it isn't there:
```sh
$ unzip -l dist/scapy-2.5.0.dev72-py3-none-any.whl | grep 'test/'
```

File: tox.ini (hunk 2 above)
Comment (evverx) [reply tox.ini:146]: Thanks for adding the test! It passes locally.

As far as I can tell it failed on the CI because the "checkout" action cloned it with `--depth=1`:
```sh
git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=1 origin +417a1525419b9db10b379aedb221cefc277c5fbc:refs/remotes/pull/3958/merge
git checkout remotes/pull/3958/merge
```
and `git_archive_id` was `'1681043547 '` instead of `'1681043522 v2.5.0-72-gedb31f4e'`. It can probably be fixed by using https://github.com/actions/checkout#fetch-all-history-for-all-tags-and-branches

File: MANIFEST.in (hunk 3 above)
Comment (gpotter2) [reply MANIFEST.in:5]: Let's ignore the warnings then and add it back.

File: doc/scapy/development.rst [hunk 4]
```diff
@@ -300,10 +300,29 @@ publish the release::
...
+Packaging Scapy
+===============
+
+When packaging Scapy, you should build the source while setting the ``SCAPY_VERSION`` variable, in order to make sure that the version remains consistent.
+
+.. code:: bash
+
+   $ SCAPY_VERSION=2.5.0 python3 -m build
+   ...
+   Successfully built scapy-2.5.0.tar.gz and scapy-2.5.0-py3-none-any.whl
```
Comment (evverx) [inline doc/scapy/development.rst:328]: I think the "scapy/VERSION" approach where it's added after unpacking the tarball still works so both Debian and Gentoo can keep using their build scripts but `SCAPY_VERSION` (or the git archive tag used by Arch and Apline) seems to be more reliable.

@thesamesam just a heads-up. `echo ${PV} > ${PN}/VERSION || die` in https://gitweb.gentoo.org/repo/gentoo.git/tree/net-analyzer/scapy/scapy-2.5.0.ebuild#n48 can be removed and replaced with `SCAPY_VERSION` when the next `scapy` version is released. It should be possible to remove it now and rely on the git archive tag though. Also the tests should be excluded from the sdist and wheel so it should be possible to remove https://gitweb.gentoo.org/repo/gentoo.git/tree/net-analyzer/scapy/files/scapy-2.5.0-no-install-tests.patch (when `scapy` with all these changes is released).

FWIW on an unrelated note looking at https://gitweb.gentoo.org/repo/gentoo.git/tree/net-analyzer/scapy/scapy-2.5.0.ebuild I think instead of removing the tests it should be possible to just skip them by passing `-K ...` to `UTscapy`. For example, `-K netaccess` should replace https://gitweb.gentoo.org/repo/gentoo.git/tree/net-analyzer/scapy/scapy-2.5.0.ebuild#n54.

File: MANIFEST.in (hunk 3 above)
Comment (evverx) [reply MANIFEST.in:5]: Makes sense. I suspect it's a setuptools issue but I have to admit I don't know what it was exactly.

File: doc/scapy/development.rst (hunk 4 above)
Comment (gpotter2) [reply doc/scapy/development.rst:328]: A note on that: the idea of [`test/run_tests`](https://github.com/secdev/scapy/blob/master/test/run_tests) is to have a script called by package maintainers that runs tests that don't require external dependencies, and not having each maintainer chose which tests to run. Might be worth adding some doc.

Comment (evverx) [reply doc/scapy/development.rst:328]: @gpotter2 I think `run_tests` uses `tox` to create a new test virtual environment and some distros can't do that so they have to run `UTscapy` directlly.

File: doc/scapy/installation.rst [hunk 5]
```diff
@@ -297,28 +287,23 @@ Solaris / SunOS requires ``libpcap`` (installed by default) to work.
 Windows
 -------
 
-.. sectionauthor:: Dirk Loss <mail at dirk-loss.de>
+You need to install Npcap order to install Scapy on Windows (should also work with Winpcap, but unsupported nowadays):
```
Comment (evverx) [inline doc/scapy/installation.rst:290]:
```suggestion
You need to install Npcap in order to install Scapy on Windows (should also work with Winpcap, but unsupported nowadays):
```

**Review by evverx — COMMENTED** [review]
LGTM. Thanks!

I think the tests now cover all corner cases (including the git archive) but just in case I created a git archive, unpacked it, run the tests from inside that directory, run `python3 -m build` with `SCAPY_VERSION` set, made sure that the test directory wasn't there, installed  the sdist (and checked `scapy.VERSION`) and then installed the wheel (and also checked `scapy.VERSION`). It all works perfectly. Thanks!

File: doc/scapy/development.rst (hunk 4 above)
Comment (evverx) [reply doc/scapy/development.rst:328]: FWIW it should be possible to run `./run_tests` in environments where `tox` isn't allowed to install dependencies with pip. To get the tests to run using the Fedora infrastructure I put the test dependencies in "BuildRequires", installed https://pypi.org/project/tox-current-env/ (Fedora ships it as a usual package) and added a sed command adding `--current-env` to `.config/ci/test.sh` to prevent `tox` from installing stuff with `pip`: https://github.com/evverx/scapy/pull/2/files#diff-0673ea92d72beb0d23ba537e1e83854385c56c702ef9abbed890b4a31ed4ece0R20

It's kludgy but it works :-) Running `UTscapy` directly is kind of hard because it already expects for example the openssl config to be generated and it's just easier to let `.config/ci/test.sh` do it.

Anyway thanks for updating the documentation! I think this thread can probably be resolved.

Comment (thesamesam) [reply doc/scapy/development.rst:328]: Thanks a bunch for the heads up @evverx!

**Comment by gpotter2** [comment]
@p-l- @guedou Hi, this ready for review+merge. Thanks !

**Comment by p-l-** [comment]
Nice work here! I'll let you merge that @gpotter2!

**Comment by MrMino** [comment]
@gpotter2 Are you planning on releasing `.whl`s to PyPI? If this PR works, then it's a 0-effort improvement for everyone, since `pip` would no longer need to build the wheel during installs (nowadays pip always installs from a wheel, even if it has to generate one first).

**Comment by evverx** [comment]
I think https://github.com/secdev/scapy/pull/3958/files#diff-7b9d684b62ceb5912e396a6250cf5c39a9fe913b0e52b5dbbd11a6d36142a23fR303-R316 kind of implies that wheels are built and uploaded when scapy is released.

**Comment by MrMino** [comment]
@evverx oh, I missed that one, thank you. All fine then.

**Comment by guedou** [comment]
Does that mean that the current release workflow described at https://scapy.readthedocs.io/en/latest/development.html will also upload the wheel using twine?

**Comment by gpotter2** [comment]
@guedou yup ! twine should upload the wheels as well, as long as they're built. This should be the case with the new build method
